File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第10章：傅里叶变换/10.1/fft2.py
# -*- coding: utf-8 -*-
import sys
import numpy as np
import cv2
#快速傅里叶变换
def fft2Image(src):
    #得到行、列
    r,c = src.shape[:2]
    #得到快速傅里叶变换最优扩充
    rPadded = cv2.getOptimalDFTSize(r)
    cPadded = cv2.getOptimalDFTSize(c)
    #边缘扩充，下边缘和右边缘扩充值为零
    fft2 = np.zeros((rPadded,cPadded,2),np.float32)
    fft2[:r,:c,0]=src
    #快速傅里叶变换
    cv2.dft(fft2,fft2,cv2.DFT_COMPLEX_OUTPUT)
    return fft2
#主函数
if __name__ =="__main__":
    if len(sys.argv) > 1:
        image = cv2.imread(sys.argv[1],cv2.CV_LOAD_IMAGE_GRAYSCALE)
    else:
        print "Usge:python fft2.py imageFile"
    #计算图像矩阵的快速傅里叶变换
    fft2 = fft2Image(image)
    #傅里叶逆变换
    ifft2 = np.zeros(fft2.shape[:2],np.float32)
    cv2.dft(fft2,ifft2,cv2.DFT_REAL_OUTPUT + cv2.DFT_INVERSE + cv2.DFT_SCALE)
    #裁剪
    img = np.copy(ifft2[:image.shape[0],:image.shape[1]])
    #裁剪后的结果 img 等于 image，两个相减的最大值为零
    print np.max(image - img)

File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第10章：傅里叶变换/10.1/fft2.cpp
#include<opencv2/core/core.hpp>
#include<opencv2/highgui/highgui.hpp>
#include<opencv2/imgproc/imgproc.hpp>
using namespace cv;
//快速傅里叶变换
void fft2Image(InputArray _src, OutputArray _dst)
{
	//得到Mat类型
	Mat src = _src.getMat();
	//判断位深
	CV_Assert(src.type() == CV_32FC1 || src.type() == CV_64FC1);
	CV_Assert(src.channels() == 1 || src.channels() == 2);
	int rows = src.rows;
	int cols = src.cols;
	//为了进行快速的傅里叶变换，我们经行和列的扩充,找到最合适扩充值
	Mat padded;
	int rPadded = getOptimalDFTSize(rows);
	int cPadded = getOptimalDFTSize(cols);
	//进行边缘扩充,扩充值为零
	copyMakeBorder(src, padded, 0, rPadded - rows, 0, cPadded - cols, BORDER_CONSTANT, Scalar::all(0));
	//快速的傅里叶变换（双通道：用于存储实部 和 虚部）
	dft(padded, _dst, DFT_COMPLEX_OUTPUT);
}
int main(int argc, char*argv[])
{
	//输入图像矩阵
	Mat img = imread(argv[1], CV_LOAD_IMAGE_GRAYSCALE);
	if (!img.data)
		return -1;
	//数据类型转换：转换为浮点型
	Mat fImg;
	img.convertTo(fImg, CV_64FC1);
	//快速傅里叶变换
	Mat fft2;
	fft2Image(fImg, fft2);
	//傅里叶逆变换
	Mat image;
	cv::dft(fft2, image, DFT_INVERSE + DFT_REAL_OUTPUT + DFT_SCALE);
	//裁剪傅里叶逆变换
	image = image(Rect(0, 0, img.cols, img.rows));
	return 0;
}
